File: csmal/csmal/Step0Repl.cs
using System;
using Deveel;

namespace csmal
{
    public class Step0Repl
    {
        private static void Main(string[] args)
        {
            while (true)
            {
                var line = Readline.ReadLine("REPL> ");
                if (line == null)
                {
                    break;
                }
                Console.WriteLine(Rep(line));
            }
        }

        private static string Read(string arg)
        {
            return arg;
        }

        private static string Eval(string arg)
        {
            return arg;
        }

        private static string Print(string arg)
        {
            return arg;
        }

        public static string Rep(string arg)
        {
            return Print(Eval(Read(arg)));
        }
    }
}

File: grust/step0_repl.rs
extern crate rustyline;

use rustyline::error::ReadlineError;
use rustyline::Editor;

fn READ(input: &str) -> &str{
    input
}

fn EVAL(form: &str) -> &str {
    form
}

fn PRINT(form: &str) -> &str {
    form
}

fn rep(line: &str) -> &str {
    PRINT(EVAL(READ(line)))
}

fn main() {
    // `()` can be used when no completer is required
    let mut rl = Editor::<()>::new();
    if rl.load_history(".mal-history").is_err() {
        eprintln!("No previous history.");
    }

    loop {
        let readline = rl.readline("user> ");
        match readline {
            Ok(line) => {
                rl.add_history_entry(&line);
                rl.save_history(".mal-history").unwrap();
                if line.len() > 0 {
                    println!("{}", rep(&line.as_ref()));
                }
            }
            Err(ReadlineError::Interrupted) => continue,
            Err(ReadlineError::Eof) => break,
            Err(err) => {
                println!("Error: {:?}", err);
                break;
            }
        }
    }
}
